akvo/rest/views/indicator_period.py

```diff
@@ -27,3 +32,23 @@ class IndicatorPeriodFrameworkViewSet(PublicProjectViewSet):
         'disaggregation_targets')
     serializer_class = IndicatorPeriodFrameworkSerializer
     project_relation = 'indicator__result__project__'
+
+
+@api_view(['POST'])
+@permission_classes((IsAuthenticated, ))
+@authentication_classes([SessionAuthentication, TastyTokenAuthentication])
+def set_periods_locked(request):
+    period_ids = request.data.get('periods', [])
+    locked = request.data.get('locked', None)
+    if len(period_ids) < 1 or locked is None:
+        return HttpResponseBadRequest()
+
+    user = request.user
+    periods = IndicatorPeriod.objects.filter(id__in=period_ids)
+
+    # only do the operation on periods where user has correct permission.
+    permitted_ids = [p.id for p in periods if user.has_perm('rsr.change_indicatorperiod', p)]
```

Comment:
This can be potentially very slow, with large projects. 

Currently, it is good enough to support indicators under a single project, and I think we should just enforce that with the API and check if the user has permissions to change indicator periods on that project. 

The permissions code itself can use some improvements to avoid doing these kind of checks hitting the DB, if we can, kind of like the recent optimisations I did for editable flags on projects, but we should think of a way to do this cleanly without adding a hackish optimisation on all such views.